File: examples/todo_app/java_todo/src/main/java/io/jamharah/todo/TodoBridge.java
/* SPDX-License-Identifier: Apache-2.0 */
package io.jamharah.todo;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonNull;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import java.util.ArrayList;
import java.util.List;
import java.util.function.BiConsumer;

/**
 * IPC dispatch + JSON (port of {@code nim_todo/todo_bridge.nim} / {@code rust_todo/src/todo_bridge.rs}).
 */
public final class TodoBridge {

  private static final Gson GSON = new Gson();

  private TodoBridge() {}

  private static void emitErr(BiConsumer<String, String> emit, int code, String msg) {
    JsonObject o = new JsonObject();
    o.addProperty("code", code);
    o.addProperty("message", msg);
    emit.accept("error", GSON.toJson(o));
  }

  private static String jStr(JsonObject pl, String k) {
    if (pl == null || !pl.has(k)) {
      return "";
    }
    JsonElement e = pl.get(k);
    return e != null && e.isJsonPrimitive() && e.getAsJsonPrimitive().isString() ? e.getAsString() : "";
  }

  private static String jStrOr(JsonObject pl, String k, String def) {
    String v = jStr(pl, k);
    return !v.isEmpty() ? v : def;
  }

  private static long jInt64(JsonObject pl, String k) {
    if (pl == null || !pl.has(k)) {
      return 0;
    }
    JsonElement e = pl.get(k);
    if (e == null || e.isJsonNull()) {
      return 0;
    }
    try {
      if (e.isJsonPrimitive()) {
        if (e.getAsJsonPrimitive().isNumber()) {
          return e.getAsLong();
        }
        if (e.getAsJsonPrimitive().isString()) {
          String s = e.getAsString();
          if (s.isEmpty()) {
            return 0;
          }
          return Long.parseLong(s);
        }
      }
    } catch (NumberFormatException ignored) {
      return 0;
    }
    return 0;
  }

  private static JsonObject categoryRowJson(TodoStoreCategoryRow r) {
    JsonObject o = new JsonObject();
    o.addProperty("id", r.id);
    o.addProperty("name", r.name);
    o.addProperty("color", r.color);
    return o;
  }

  private static JsonObject todoRowJson(TodoStoreTodoRow r) {
    JsonObject o = new JsonObject();
    o.addProperty("id", r.id);
    o.addProperty("title", r.title);
    o.addProperty("description", r.description);
    o.addProperty("status", r.status);
    o.addProperty("priority", r.priority);
    if (r.categoryIdIsNull) {
      o.add("categoryId", JsonNull.INSTANCE);
    } else {
      o.addProperty("categoryId", r.categoryId);
    }
    if (r.parentIdIsNull) {
      o.add("parentId", JsonNull.INSTANCE);
    } else {
      o.addProperty("parentId", r.parentId);
    }
    if (r.dueDateIsNull) {
      o.add("dueDate", JsonNull.INSTANCE);
    } else {
      o.addProperty("dueDate", r.dueDate);
    }
    o.addProperty("position", r.position);
    o.addProperty("createdAt", r.createdAt);
    o.addProperty("updatedAt", r.updatedAt);
    return o;
  }

  private static int emitCategoriesList(TodoStore s, BiConsumer<String, String> emit) {
    List<TodoStoreCategoryRow> rows = new ArrayList<>();
    int rc = s.storeCategoriesForeach(rows::add);
    if (rc != SqliteRc.SQLITE_OK) {
      emitErr(emit, rc, "categories.foreach failed");
      return -1;
    }
    JsonArray arr = new JsonArray();
    for (TodoStoreCategoryRow r : rows) {
      arr.add(categoryRowJson(r));
    }
    emit.accept("categories.list", GSON.toJson(arr));
    return 0;
  }

  private static int emitTodosList(TodoStore s, BiConsumer<String, String> emit) {
    List<TodoStoreTodoRow> rows = new ArrayList<>();
    int rc = s.storeTodosForeach(rows::add);
    if (rc != SqliteRc.SQLITE_OK) {
      emitErr(emit, rc, "todos.foreach failed");
      return -1;
    }
    JsonArray arr = new JsonArray();
    for (TodoStoreTodoRow r : rows) {
      arr.add(todoRowJson(r));
    }
    emit.accept("todos.list", GSON.toJson(arr));
    return 0;
  }

  /** Handles one {@code __nv} wire JSON envelope. Returns {@code 0} on success, {@code -1} on failure. */
  public static int bridgeHandleWire(TodoStore s, String rawIpcJson, BiConsumer<String, String> emit) {
    if (s == null || emit == null || rawIpcJson == null || rawIpcJson.isEmpty()) {
      return -1;
    }
    JsonObject root;
    try {
      JsonElement el = com.google.gson.JsonParser.parseString(rawIpcJson);
      root = el != null && el.isJsonObject() ? el.getAsJsonObject() : null;
    } catch (JsonParseException e) {
      emitErr(emit, -1, "invalid wire json");
      return -1;
    }
    if (root == null) {
      emitErr(emit, -1, "invalid wire json");
      return -1;
    }
    if (!root.has("d") || !root.get("d").isJsonObject()) {
      emitErr(emit, -1, "missing wire.d object");
      return -1;
    }
    JsonObject d = root.getAsJsonObject("d");
    String action = "";
    if (d.has("action") && d.get("action").isJsonPrimitive()) {
      action = d.get("action").getAsString();
    }
    if (action.isEmpty()) {
      emitErr(emit, -1, "missing action");
      return -1;
    }
    JsonObject pl = null;
    if (d.has("payload") && d.get("payload").isJsonObject()) {
      pl = d.getAsJsonObject("payload");
    }

    if ("categories.list".equals(action)) {
      return emitCategoriesList(s, emit);
    }
    if ("todos.list".equals(action)) {
      return emitTodosList(s, emit);
    }
    if ("categories.create".equals(action)) {
      if (pl == null) {
        emitErr(emit, SqliteRc.SQLITE_MISUSE, "missing payload");
        return -1;
      }
      String name = jStr(pl, "name");
      String color = jStrOr(pl, "color", "#6366f1");
      long[] newId = new long[1];
      int rc = s.storeCategoryInsert(name, color, newId);
      if (rc != SqliteRc.SQLITE_OK) {
        emitErr(emit, rc, "category insert failed");
        return -1;
      }
      return emitCategoriesList(s, emit);
    }
    if ("categories.delete".equals(action)) {
      if (pl == null) {
        emitErr(emit, SqliteRc.SQLITE_MISUSE, "missing payload");
        return -1;
      }
      long id = jInt64(pl, "id");
      int rc = s.storeCategoryDeleteById(id);
      if (rc != SqliteRc.SQLITE_OK) {
        emitErr(emit, rc, "category delete failed");
        return -1;
      }
      return emitCategoriesList(s, emit);
    }
    if ("todos.create".equals(action)) {
      if (pl == null) {
        emitErr(emit, SqliteRc.SQLITE_MISUSE, "missing payload");
        return -1;
      }
      String title = jStr(pl, "title");
      String desc = jStrOr(pl, "description", "");
      String status = jStrOr(pl, "status", "pending");
      String pri = jStrOr(pl, "priority", "medium");
      long cat = jInt64(pl, "categoryId");
      long par = jInt64(pl, "parentId");
      long due = jInt64(pl, "dueDate");
      boolean catSet = cat > 0;
      boolean parSet = par > 0;
      boolean dueSet = due != 0;
      long[] newId = new long[1];
      int rc =
          s.storeTodoInsert(title, desc, status, pri, catSet, cat, parSet, par, dueSet, due, newId);
      if (rc != SqliteRc.SQLITE_OK) {
        emitErr(emit, rc, "todo insert failed");
        return -1;
      }
      return emitTodosList(s, emit);
    }
    if ("todos.update".equals(action)) {
      if (pl == null) {
        emitErr(emit, SqliteRc.SQLITE_MISUSE, "missing payload");
        return -1;
      }
      long id = jInt64(pl, "id");
      String title = jStr(pl, "title");
      String desc = jStrOr(pl, "description", "");
      String status = jStrOr(pl, "status", "pending");
      String pri = jStrOr(pl, "priority", "medium");
      long cat = jInt64(pl, "categoryId");
      long par = jInt64(pl, "parentId");
      long due = jInt64(pl, "dueDate");
      int pos = (int) jInt64(pl, "position");
      boolean catSet = cat > 0;
      boolean parSet = par > 0;
      boolean dueSet = due != 0;
      int rc =
          s.storeTodoUpdateFull(id, title, desc, status, pri, catSet, cat, parSet, par, dueSet, due, pos);
      if (rc != SqliteRc.SQLITE_OK) {
        emitErr(emit, rc, "todo update failed");
        return -1;
      }
      return emitTodosList(s, emit);
    }
    if ("todos.delete".equals(action)) {
      if (pl == null) {
        emitErr(emit, SqliteRc.SQLITE_MISUSE, "missing payload");
        return -1;
      }
      long id = jInt64(pl, "id");
      int rc = s.storeTodoDeleteById(id);
      if (rc != SqliteRc.SQLITE_OK) {
        emitErr(emit, rc, "todo delete failed");
        return -1;
      }
      return emitTodosList(s, emit);
    }
    emitErr(emit, -1, "unknown action");
    return -1;
  }
}

File: examples/todo_app/py_todo/todo_bridge.py
# IPC dispatch + JSON (port of c_todo/src/todo_bridge.c / nim_todo/todo_bridge.nim).
# SPDX-License-Identifier: Apache-2.0

from __future__ import annotations

import json
from collections.abc import Callable
from typing import Any

from todo_store import (
    SQLITE_MISUSE,
    SQLITE_OK,
    TodoStore,
    TodoStoreCategoryRow,
    TodoStoreTodoRow,
    store_categories_foreach,
    store_category_delete_by_id,
    store_category_insert,
    store_todo_delete_by_id,
    store_todo_insert,
    store_todos_foreach,
    store_todo_update_full,
)


def _emit_err(emit: Callable[[str, str], None], code: int, msg: str) -> None:
    emit("error", json.dumps({"code": code, "message": msg}))


def _j_str(pl: dict[str, Any] | None, k: str) -> str:
    if not pl or k not in pl:
        return ""
    v = pl[k]
    if isinstance(v, str):
        return v
    return ""


def _j_str_or(pl: dict[str, Any] | None, k: str, default: str) -> str:
    s = _j_str(pl, k)
    return s if s else default


def _j_int64(pl: dict[str, Any] | None, k: str) -> int:
    if not pl or k not in pl:
        return 0
    v = pl[k]
    if isinstance(v, bool):
        return int(v)
    if isinstance(v, int):
        return int(v)
    return 0


def _category_row_json(r: TodoStoreCategoryRow) -> dict[str, Any]:
    return {"id": r.id, "name": r.name, "color": r.color}


def _todo_row_json(r: TodoStoreTodoRow) -> dict[str, Any]:
    o: dict[str, Any] = {
        "id": r.id,
        "title": r.title,
        "description": r.description,
        "status": r.status,
        "priority": r.priority,
        "position": r.position,
        "createdAt": r.created_at,
        "updatedAt": r.updated_at,
    }
    if r.category_id_is_null:
        o["categoryId"] = None
    else:
        o["categoryId"] = r.category_id
    if r.parent_id_is_null:
        o["parentId"] = None
    else:
        o["parentId"] = r.parent_id
    if r.due_date_is_null:
        o["dueDate"] = None
    else:
        o["dueDate"] = r.due_date
    return o


def _emit_categories_list(s: TodoStore, emit: Callable[[str, str], None]) -> int:
    rows: list[TodoStoreCategoryRow] = []

    def _cb(r: TodoStoreCategoryRow) -> None:
        rows.append(r)

    rc = store_categories_foreach(s, _cb)
    if rc != SQLITE_OK:
        _emit_err(emit, rc, "categories.foreach failed")
        return -1
    emit("categories.list", json.dumps([_category_row_json(r) for r in rows]))
    return 0


def _emit_todos_list(s: TodoStore, emit: Callable[[str, str], None]) -> int:
    rows: list[TodoStoreTodoRow] = []

    def _cb(r: TodoStoreTodoRow) -> None:
        rows.append(r)

    rc = store_todos_foreach(s, _cb)
    if rc != SQLITE_OK:
        _emit_err(emit, rc, "todos.foreach failed")
        return -1
    emit("todos.list", json.dumps([_todo_row_json(r) for r in rows]))
    return 0


def bridge_handle_wire(
    s: TodoStore | None,
    raw_ipc_json: str,
    emit: Callable[[str, str], None],
) -> int:
    if s is None or not raw_ipc_json:
        return -1
    try:
        root = json.loads(raw_ipc_json)
    except json.JSONDecodeError:
        _emit_err(emit, -1, "invalid wire json")
        return -1
    if not isinstance(root, dict):
        _emit_err(emit, -1, "invalid wire json")
        return -1
    d = root.get("d")
    if not isinstance(d, dict):
        _emit_err(emit, -1, "missing wire.d object")
        return -1
    action = d.get("action", "")
    if not isinstance(action, str) or not action:
        _emit_err(emit, -1, "missing action")
        return -1
    pl = d.get("payload")
    pld: dict[str, Any] | None = pl if isinstance(pl, dict) else None

    if action == "categories.list":
        return _emit_categories_list(s, emit)

    if action == "todos.list":
        return _emit_todos_list(s, emit)

    if action == "categories.create":
        if pld is None:
            _emit_err(emit, SQLITE_MISUSE, "missing payload")
            return -1
        name = _j_str(pld, "name")
        color = _j_str_or(pld, "color", "#6366f1")
        new_id: list[int] = []
        rc = store_category_insert(s, name, color, new_id)
        if rc != SQLITE_OK:
            _emit_err(emit, rc, "category insert failed")
            return -1
        return _emit_categories_list(s, emit)

    if action == "categories.delete":
        if pld is None:
            _emit_err(emit, SQLITE_MISUSE, "missing payload")
            return -1
        row_id = _j_int64(pld, "id")
        rc = store_category_delete_by_id(s, row_id)
        if rc != SQLITE_OK:
            _emit_err(emit, rc, "category delete failed")
            return -1
        return _emit_categories_list(s, emit)

    if action == "todos.create":
        if pld is None:
            _emit_err(emit, SQLITE_MISUSE, "missing payload")
            return -1
        title = _j_str(pld, "title")
        desc = _j_str_or(pld, "description", "")
        status = _j_str_or(pld, "status", "pending")
        pri = _j_str_or(pld, "priority", "medium")
        cat = _j_int64(pld, "categoryId")
        par = _j_int64(pld, "parentId")
        due = _j_int64(pld, "dueDate")
        cat_set = cat > 0
        par_set = par > 0
        due_set = due != 0
        new_id: list[int] = []
        rc = store_todo_insert(
            s, title, desc, status, pri, cat_set, cat, par_set, par, due_set, due, new_id
        )
        if rc != SQLITE_OK:
            _emit_err(emit, rc, "todo insert failed")
            return -1
        return _emit_todos_list(s, emit)

    if action == "todos.update":
        if pld is None:
            _emit_err(emit, SQLITE_MISUSE, "missing payload")
            return -1
        row_id = _j_int64(pld, "id")
        title = _j_str(pld, "title")
        desc = _j_str_or(pld, "description", "")
        status = _j_str_or(pld, "status", "pending")
        pri = _j_str_or(pld, "priority", "medium")
        cat = _j_int64(pld, "categoryId")
        par = _j_int64(pld, "parentId")
        due = _j_int64(pld, "dueDate")
        pos = int(_j_int64(pld, "position"))
        cat_set = cat > 0
        par_set = par > 0
        due_set = due != 0
        rc = store_todo_update_full(
            s, row_id, title, desc, status, pri, cat_set, cat, par_set, par, due_set, due, pos
        )
        if rc != SQLITE_OK:
            _emit_err(emit, rc, "todo update failed")
            return -1
        return _emit_todos_list(s, emit)

    if action == "todos.delete":
        if pld is None:
            _emit_err(emit, SQLITE_MISUSE, "missing payload")
            return -1
        row_id = _j_int64(pld, "id")
        rc = store_todo_delete_by_id(s, row_id)
        if rc != SQLITE_OK:
            _emit_err(emit, rc, "todo delete failed")
            return -1
        return _emit_todos_list(s, emit)

    _emit_err(emit, -1, "unknown action")
    return -1
